central/complianceoperator/v2/report/manager/watcher/scanwatcher.go

```diff
@@ -0,0 +1,239 @@
+// GetWatcherIDFromScan given a Scan, returns a unique ID for the watcher
+func GetWatcherIDFromScan(scan *storage.ComplianceOperatorScanV2) (string, error) {
+	clusterID := scan.GetClusterId()
+	if clusterID == "" {
+		return "", errors.New("Missing cluster ID")
+	}
+	scanID := scan.GetId()
+	if scanID == "" {
+		return "", errors.New("Missing scan ID")
+	}
+	startTime := scan.GetLastStartedTime()
+	if startTime == nil {
+		// This could happen if the scan is freshly created
+		log.Debug("Missing last stated time")
+		return "", nil
+	}
+	return fmt.Sprintf("%s:%s:%s", clusterID, scanID, startTime.String()), nil
+}
+
+// GetWatcherIDFromCheckResult given a CheckResult, returns a unique ID for the watcher
+func GetWatcherIDFromCheckResult(ctx context.Context, result *storage.ComplianceOperatorCheckResultV2, scanDataStore scanDS.DataStore) (string, error) {
+	scanRefQuery := search.NewQueryBuilder().AddExactMatches(
+		search.ComplianceOperatorScanRef,
+		result.GetScanRefId(),
+	).ProtoQuery()
+	scans, err := scanDataStore.SearchScans(ctx, scanRefQuery)
+	if err != nil {
+		return "", errors.Errorf("Unable to retrieve scan : %v", err)
+	}
+	if len(scans) == 0 {
+		return "", errors.Errorf("Scan not found for result %s", result.GetCheckName())
+	}
+	var startTime string
+	var ok bool
+	if startTime, ok = result.GetAnnotations()[lastScannedAnnotationKey]; !ok {
+		return "", errors.Errorf("%s annotation not found", lastScannedAnnotationKey)
+	}
+	timestamp, err := protocompat.ParseRFC3339NanoTimestamp(startTime)
+	if err != nil {
+		return "", errors.Errorf("Unable to parse time: %v", err)
+	}
+	if timestamp.String() != scans[0].GetLastStartedTime().String() {
+		return "", errors.New("The result and the scan do not have the same timestamp")
+	}
+	return GetWatcherIDFromScan(scans[0])
+}
+
+// readyQueue represents the expected queue interface to push the results
+type readyQueue[T comparable] interface {
+	Push(T)
+}
+
+type scanWatcherImpl struct {
+	ctx     context.Context
+	cancel  func()
+	scanC   chan *storage.ComplianceOperatorScanV2
+	resultC chan *storage.ComplianceOperatorCheckResultV2
+	stopped *concurrency.Signal
+
+	readyQueue  readyQueue[*ScanWatcherResults]
+	scanResults *ScanWatcherResults
+	totalChecks int
+}
+
+// NewScanWatcher creates a new ScanWatcher
+func NewScanWatcher(ctx context.Context, watcherID string, queue readyQueue[*ScanWatcherResults]) *scanWatcherImpl {
+	watcherCtx, cancel := context.WithCancel(ctx)
+	finishedSignal := concurrency.NewSignal()
+	ret := &scanWatcherImpl{
+		ctx:        watcherCtx,
+		cancel:     cancel,
+		scanC:      make(chan *storage.ComplianceOperatorScanV2, defaultChanelSize),
+		resultC:    make(chan *storage.ComplianceOperatorCheckResultV2, defaultChanelSize),
+		stopped:    &finishedSignal,
+		readyQueue: queue,
+		scanResults: &ScanWatcherResults{
+			Ctx:          ctx,
+			WatcherID:    watcherID,
+			CheckResults: set.NewStringSet(),
+		},
+	}
+	timeout := time.NewTimer(defaultTimeout)
```

Comment:
> Before Go 1.23, the garbage collector did not recover timers that had not yet expired or been stopped, so code often immediately deferred t.Stop after calling NewTimer, to make the timer recoverable when it was no longer needed. As of Go 1.23, the garbage collector can recover unreferenced timers, even if they haven't expired or been stopped. The Stop method is no longer necessary to help the garbage collector. (Code may of course still want to call Stop to stop the timer for other reasons.)

https://pkg.go.dev/time#NewTimer

Maybe we need that defer? We are not yet on 1.23